### PR TITLE
add regexp.prototype.flags

### PR DESCRIPTION
**Pull request by ljharb**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:

- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by ljharb** [comment]
@jakebailey i'd love some assistance here if you or anyone else has the time :-)

**Comment by jakebailey** [comment]
If you really want to use a ridiculously large template literal type like this, you can only copy and paste all of the output into the expected type. The rule matches strings, it can't evaluate anything.

That or don't use ExpectType for this; there are other ways to test for correct types using assignability, for example.

**Comment by jakebailey** [comment]
That or give this type a name instead of inlining it at the return, which is probably nicer anyway.

File: types/regexp.prototype.flags/regexp.prototype.flags-tests.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import "regexp.prototype.flags/auto";
+
+type D = 'd' | '';
+type G = 'g' | '';
+type I = 'i' | '';
+type M = 'm' | '';
+type S = 's' | '';
+type U = 'u' | '';
+type V = 'v' | '';
+type Y = 'y' | '';
```
Comment (jakebailey) [inline types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:11]: This all won't really work in a test. I would just name your type in implementation.d.ts.

Comment (ljharb) [reply types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:11]: hmm, for CJS compat i need `export =`, so i can't export a separate type. i tried to get at it with `ReturnType<>` but it doesn't work.

Comment (jakebailey) [reply types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:11]: You can using:

```ts
namespace foo {
    type Flags = ...;
}

function foo(): foo.Flags;

export = foo;
```

Comment (ljharb) [reply types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:11]: hm, ok - and on the consuming side, i can `import type { Flags } from` it?

Comment (jakebailey) [reply types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:11]: There's an example in the readme; I'd link it but I'm on my phone :P

Comment (jakebailey) [reply types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:11]: Yeah, downstream with esModuleInterop, but on DT tests you'd do a require and then fully qualify it.

Comment (ljharb) [reply types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:11]: thanks! i'll see what i can do.

Comment (ljharb) [reply types/regexp.prototype.flags/regexp.prototype.flags-tests.ts]: this seems to work

File: types/regexp.prototype.flags/regexp.prototype.flags-tests.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import flags = require("regexp.prototype.flags");
+import "regexp.prototype.flags/auto";
+
+flags(/foo/) === ""; // $ExpectType boolean
```
Comment (jakebailey) [inline types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:4]: Hm, did this not work as just Flags without the comparison? Any comparison is going to be boolean, unless I'm missing something...

Comment (ljharb) [reply types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:4]: nope, sadly it didn't :-/ it said, basically, "expected foo.Flags, got <the actual full union type foo.Flags represents>" which was the same kind of error i had before.

it's very weird to me that the tests wouldn't either expand `foo.Flags` to the full union, or, realize the full union is the same as `foo.Flags` and pass.

Comment (jakebailey) [reply types/regexp.prototype.flags/regexp.prototype.flags-tests.ts:4]: Weird, I have no idea why the alias isn't preserved in this case, but I also see that locally.

**Comment by jakebailey** [comment]
In trying to come up with a better test, I noticed that you can't actually import the `Flags` type from the public API. I realized that you were exporting a const at the index, but I think this may acutally be better:

```
import implementation = require("./implementation");
import polyfill = require("./polyfill");
import shim = require("./shim");

type Implementation = typeof implementation;
type Polyfill = typeof polyfill;
type Shim = typeof shim;

declare namespace index {
    type Flags = implementation.Flags;

    const implementation: Implementation;
    const polyfill: Polyfill;
    const shim: Shim;
}

declare function index(re: ThisParameterType<Implementation>): implementation.Flags;


export = index;
```

Then the test can be:

```ts
import flags = require("regexp.prototype.flags");
import "regexp.prototype.flags/auto";

declare function isFlags(f: flags.Flags): void;

isFlags(flags(/foo/));
isFlags(flags(/foo/g));
isFlags(flags(/foo/i));
isFlags(flags(/foo/m));
isFlags(flags(/foo/s));
isFlags(flags(/foo/u));
isFlags(flags(/foo/v));
isFlags(flags(/foo/y));
```

I guess I don't recall how you have typed other packages of this format to know if they suffer from the same problem, though. Up to you.

**Comment by jakebailey** [comment]
(effectively, using namespaces or real imports/exports are better because they preserve the ability to access types from the imported symbol)

**Comment by ljharb** [comment]
that looks like a much better solution! i've never used the namespace approach before, and the types for things have worked fine so far (but were much simpler than this one), but i like this approach.

**Comment by jakebailey** [comment]
```ts
export import Flags = implementation.Flags;
```

Theoretically also works, but needs https://github.com/microsoft/DefinitelyTyped-tools/pull/966 as it produces the wrong lint. The alias is also fine (just half indirect).

**Review by jakebailey — APPROVED** [review]

**Comment by ljharb** [comment]
Ready to merge

**Comment by ljharb** [comment]
Thanks for your help!

**Pull request merged**